Tests and documentation for Platform Package installation routines (#494)
heroku/heroku-buildpack-php: support/installer/README.md

- `$STACK` for the name of the current stack ("heroku-20" etc)

First, it generates a list of all repositories that should be used. Packagist gets disabled as a repository (we do not want to install public packages here), then a `path` type repository (forcing copying vs symlinking, due to different filesystems at build time and runtime) points Composer to our installer plugin (given to the program as the first argument).

Next, it lists all `composer` type repositories - that's usually the default platform repository, plus any other custom repositories given by the user, and they're given as arguments to the. These are given as additional arguments to the program.

For each package listed as a dependency in the project's `composer.lock`, it then writes a metapackage of the same name and version to a list of `path` `repositories` in "`platform.json`". For each platform package entries in the original package's `require` section, it rewrites that entry with a prefix of "heroku-sys/", so if a package requires "ext-mbstring", it writes "heroku-sys/ext-mbstring" into the metapackage's `require` section instead.

> To this same list, it also writes a metapackage named `composer.json/composer.lock` (this makes for nice output in the case of package resolution failures) containing all of the project's direct platform package `require`s from `composer.json`, and places that package into the `require` section of "`platform.json`" along with all the userland dependency metapackages from the paragraph above, as well as a few default platform packages such as `heroku-sys/composer` (along with a require for the lock file's `composer-plugin-api` version, which will automatically ensure the correct installation of Composer 1 or Composer 2), or `heroku-sys/apache`.

Reviewer: ```suggestion
To this same list, it also writes a metapackage named `composer.json/composer.lock` (this makes for nice output in the case of package resolution failures) containing all of the project's direct platform package `require`s from `composer.json`. That package is added to the `require` section of "`platform.json`" along with all the userland dependency metapackages from the paragraph above, as well as a few default platform packages such as `heroku-sys/composer` (along with a require for the lock file's `composer-plugin-api` version, which will automatically ensure the correct installation of Composer 1 or Composer 2), or `heroku-sys/apache`.
```

> , or `heroku-sys/apache`.

Also, is nginx missing here (before the "or")?

                {
                    "type": "metapackage",
                    "name": "composer.json\/composer.lock-require-dev",
                    "version": "dev-c2b9dcae256d1b255b7265eef089f6c3",
                    "require": {
                        "heroku-sys\/ext-pcov": "*"
                    },
                    "replace": {},
                    "provide": {},
                    "conflict": {}
                }
            ]
        }
    ]
}
```

As you can see in that example, the original project's `composer.json` required a bunch of extensions, but no PHP runtime version, as well as `symfony/process`. In `require-dev`, it had `kahlan/kahlan` and `ext-pcov`, and because this was a Heroku CI install (as indicated by the `$HEROKU_PHP_INSTALL_DEV` env var), it generated metapackages for those as well.

The effective PHP runtime version that will be chosen for this install will be at least 7.2.5 (but not PHP 8), up to the maximum possible version that has all the required extensions available either bundled or as separate packages, since that's what the solver will conclude from the `symfony/process` package's requirements.

## Platform Repositories

> See `support/build/README.md` for details.

Reviewer: ```suggestion
See [support/build/README.md](../build/README.md) for details.
```
